compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -730,26 +715,246 @@ fn encode_ty<'tcx>(
+                )
+            }
+        };
+        let mut fn_abi = fn_abi.clone();
+        // HACK(rcvalle): It is okay to not replace or update the entire ArgAbi here because the
+        //   other fields are never used.
+        fn_abi.args[0].layout.ty = self_ty;
+
+        return fn_abi;
+    } else if is_closure_call(tcx, fn_abi)
+        && let ty::Closure(_, args) = fn_abi.args[0].layout.ty.peel_refs().kind()
+    {
+        let fn_sig = tcx.signature_unclosure(args.as_closure().sig(), hir::Unsafety::Normal);
+        return tcx
+            .fn_abi_of_fn_ptr(ty::ParamEnv::empty().and((fn_sig, List::empty())))
+            .unwrap_or_else(|fn_abi_error| {
+                bug!("transform_fnabi: couldn't get fn_abi of fn_sig {:?}", fn_abi_error)
+            })
+            .clone();
+    } else if is_dynamic_fn_call(tcx, fn_abi) {
+        let fn_sig = ty::Binder::bind_with_vars(
+            tcx.mk_fn_sig(
+                fn_abi.args[1..].iter().map(|arg| arg.layout.ty),
+                fn_abi.ret.layout.ty,
+                false,
+                hir::Unsafety::Normal,
+                Abi::Rust,
+            ),
+            List::empty(),
+        );
+        return tcx
+            .fn_abi_of_fn_ptr(ty::ParamEnv::empty().and((fn_sig, List::empty())))
+            .unwrap_or_else(|fn_abi_error| {
+                bug!("transform_fnabi: couldn't get fn_abi of fn_sig {:?}", fn_abi_error)
+            })
+            .clone();
+    }
+
+    fn_abi.clone()
+}
+
+/// Transforms a predicate for being encoded and used in the substitution dictionary.
+fn transform_predicate<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    predicate: ty::PolyExistentialPredicate<'tcx>,
+    _options: EncodeTyOptions,
+) -> ty::PolyExistentialPredicate<'tcx> {
+    match predicate.as_ref().skip_binder() {
+        ty::ExistentialPredicate::Trait(trait_ref) => {
+            if tcx.is_fn_trait(trait_ref.def_id) {
+                bug!("transform_predicate: unexpected Fn trait `{}`", trait_ref)
+            } else {
+                // Transform non Fn traits into their identities
+                let trait_ref = ty::TraitRef::identity(tcx, trait_ref.def_id);
+                ty::Binder::dummy(ty::ExistentialPredicate::Trait(
+                    ty::ExistentialTraitRef::erase_self_ty(tcx, trait_ref),
+                ))
+            }
+        }
+        _ => predicate,
+    }
+}
+
 /// Transforms predicates for being encoded and used in the substitution dictionary.
 fn transform_predicates<'tcx>(
     tcx: TyCtxt<'tcx>,
     predicates: &List<ty::PolyExistentialPredicate<'tcx>>,
-    _options: EncodeTyOptions,
+    options: EncodeTyOptions,
 ) -> &'tcx List<ty::PolyExistentialPredicate<'tcx>> {
     let predicates: Vec<ty::PolyExistentialPredicate<'tcx>> = predicates
+        .iter()
+        .filter_map(|predicate| Some(transform_predicate(tcx, predicate, options)))
+        .collect();
+
+    // Use either the base trait or autotraits only as discriminant
```

Comment:
I'm not sure if I understand why this is correct. That means that `dyn Trait + Auto` is not distinct from `dyn Trait`, but then `dyn Auto + Auto2` is distinct from `dyn Auto`.

Reply:
Ping

Reply:
Reverted it and enforced sorting the predicates because in old LLVM CI build the predicates ordering seemed to change. (The original removes predicates because these aren't available at call sites.)

Reply:
Sorry, removes projections, not predicates.